arrows/ffmpeg/ffmpeg_video_input.cxx

```diff
@@ -0,0 +1,832 @@
+  }
+
+  //is_->contig_memory_ = 0;
+  d->f_video_index = -1;
+  d->f_data_index = -1;
+  d->f_start_time = -1;
+  d->video_path = "";
+  d->frame_advanced = 0;
+  d->end_of_video = true;
+  d->number_of_frames = 0;
+  d->number_of_frames_valid = false;
+  //is_->metadata_.clear();
+  if (d->f_video_stream)
+  {
+    avcodec_close(d->f_video_stream ->codec);
+    d->f_video_stream = nullptr;
+  }
+  if (d->f_format_context)
+  {
+    avformat_close_input(&d->f_format_context);
+    d->f_format_context = nullptr;
+  }
+
+  d->f_video_encoding = nullptr;
+}
+
+// ------------------------------------------------------------------
+bool
+ffmpeg_video_input
+::next_frame( kwiver::vital::timestamp& ts,
+              uint32_t timeout )
+{
+  if (!d->is_opened())
+  {
+    throw vital::file_not_read_exception(d->video_path, "Video not open");
+  }
+
+  bool ret = d->advance();
+  if (!ret && !d->end_of_video)
+  {
+    // If it's the first frame that fails
+    // -> first frame as the end of the video
+    d->number_of_frames_valid = true;
+  }
+
+  d->end_of_video = !ret;
+  if (ret)
+  {
+    ts = this->frame_timestamp();
+    if (!d->number_of_frames_valid)
+    {
+      d->number_of_frames = ts.get_frame();
+    }
+  };
+  return ret;
+}
+
+// ------------------------------------------------------------------
+bool ffmpeg_video_input::seek_frame(kwiver::vital::timestamp& ts,
+  kwiver::vital::timestamp::frame_t frame_number,
+  uint32_t timeout)
+{
+  // Quick return if the stream isn't open.
+  if (!d->is_opened())
+  {
+    throw vital::file_not_read_exception(d->video_path, "Video not open");
+    return false;
+  }
+  if (frame_number <= 0)
+  {
+    return false;
+  }
+
+  // Quick return if the stream isn't open.
+  if (timeout != 0)
+  {
+    LOG_WARN(this->logger(), "Timeout argument is not supported.");
+  }
+
+  int current_frame_number = this->frame_timestamp().get_frame();
+  // If current frame number is greater than requested frame reopen
+  // file to reset to start
+  if (current_frame_number > frame_number)
+  {
+    this->open(d->video_path); // Calls close on current video
+    current_frame_number = this->frame_timestamp().get_frame();
+  }
+
+  // Just advance video until the requested frame is reached
+  for (int i = current_frame_number; i < frame_number; ++i)
+  {
+    if (!this->next_frame(ts))
+    {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+
+// ------------------------------------------------------------------
+kwiver::vital::image_container_sptr
+ffmpeg_video_input
+::frame_image( )
+{
+  // Quick return if the stream isn't valid
+  if (!d->is_valid())
+  {
+    return nullptr;
+  }
+
+  AVCodecContext* enc = d->f_format_context->streams[d->f_video_index]->codec;
+
+  // If we have not already converted this frame, try to convert it
+  if (!d->current_image_memory && d->f_frame->data[0] != 0)
+  {
+    int width = enc->width;
+    int height = enc->height;
+    int num_pixels = 3;
```

Comment:
`num_pixels` is a bit misleading for this variable name.  I would call it `num_channels` or `depth`.

Reply:
Good catch ! Solved with https://github.com/Kitware/kwiver/pull/533/commits/a1384a87ec522d79109bf6abf29f1515c2ff40a1